src/devices/MatrixKeyboard/MatrixKeyboard.cs

```diff
@@ -0,0 +1,147 @@
+            buttonValues = new PinValue[this.rowPins.Length * this.colPins.Length];
+
+            scanTimer = new Timer(1000 / (scanFreq * this.rowPins.Length));
+            scanTimer.Elapsed += ScanTimerElapsed;
+        }
+
+        /// <summary>
+        /// Open GPIO pins then start keyboard scanning
+        /// </summary>
+        public void StartScan()
+        {
+            for (var i = 0; i < rowPins.Length; i++)
+            {
+                gpio.OpenPin(rowPins[i], PinMode.Output);
+            }
+            for (var i = 0; i < colPins.Length; i++)
+            {
+                gpio.OpenPin(colPins[i], PinMode.Input);
+            }
+            scanTimer.Start();
+        }
+
+        /// <summary>
+        /// Stop keyboard scanning then close GPIO pins
+        /// </summary>
+        public void StopScan()
+        {
+            scanTimer.Stop();
+            for (var i = 0; i < rowPins.Length; i++)
+            {
+                gpio.ClosePin(rowPins[i]);
+            }
+            for (var i = 0; i < colPins.Length; i++)
+            {
+                gpio.ClosePin(colPins[i]);
+            }
+        }
+
+        /// <summary>
+        /// Get all buttons' values
+        /// </summary>
+        public ReadOnlySpan<PinValue> Values => buttonValues.AsSpan();
+
+        /// <summary>
+        /// Get buttons' values by row
+        /// </summary>
+        /// <param name="row">Row index</param>
+        public ReadOnlySpan<PinValue> RowValues(int row) => buttonValues.AsSpan(row * rowPins.Length, colPins.Length);
+
+        /// <summary>
+        /// Keyboard event
+        /// </summary>
+        public event PinChangeEventHandler PinChangeEvent;
+
+        /// <summary>
+        /// Delegate of keyboard event
+        /// </summary>
+        public delegate void PinChangeEventHandler(object sender, MatrixKeyboardEventArgs pinValueChangedEventArgs);
+
+        public void Dispose()
+        {
+            scanTimer?.Dispose();
+            scanTimer = null;
+
+            gpio?.Dispose();
+            gpio = null;
+        }
+
+        private readonly int[] rowPins;
+        private readonly int[] colPins;
+        private int currentRow;
+        private readonly PinValue[] buttonValues;
+
+        private IGpioController gpio;
+        private Timer scanTimer;
+        private void ScanTimerElapsed(object sender, ElapsedEventArgs e)
+        {
+            for (var i = 0; i < colPins.Length; i++)
+            {
+                var index = currentRow * colPins.Length + i;
+
+                var oldValue = buttonValues[index];
+                var newValue = gpio.Read(colPins[i]);
+
+                buttonValues[index] = newValue;
+                if (newValue != oldValue)
+                {
+                    var args = new MatrixKeyboardEventArgs(newValue == PinValue.High ? PinEventTypes.Rising : PinEventTypes.Falling, currentRow, i);
+                    PinChangeEvent(this, args);
```

Comment:
Why are you manually doing your eventing model when we have eventing APIS in the GpioController? Did you try using `controller.RegisterCallbackForPinValueChangedEvent`

Reply:
nope, `controller.RegisterCallbackForPinValueChangedEvent` is for `GpioController`, but I'm using `IGpioController` to compatible with all GPIO controllers like `MCP23017`. `IGpioController` doesn't have any event I can use